### PR TITLE
chore: inline pces proposal

### LINKED ISSUE #13777: Write proposal for elimination of accidental self branches
It's possible for an honest node to create a branch in the hashgraph if it creates a self event, gossips that event, and then crashes prior to that event being written to a PCES file.

Propose a solution that makes it impossible for a node to self-branch in this way.

### PR DESCRIPTION
**Pull request by cody-littley**
closes #13777

### CODE REVIEW

File: platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+time, the node will build a new self-event on the same parent as the previous self-event (or with no self parent),
+causing a branch.
+
+This happens today most frequently on update boundaries when **every node** deterministically shutdown and then
+start back up. In theory it can also happen when individual nodes are restarted.
+
+### Requirements
+
+1. There must not be any noticeable impact in time-to-consensus.
+    a) if there is an observed impact, all product stakeholders must agree that the impact is negligible
```
Comment (hendrikebbers) [inline platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md:49]: Why a) if there is no b). Next to that the formatting is broken: you need to add a newline to have it formatted correctly in MD.

**Review by hendrikebbers — COMMENTED** [review]
Started between 2 events, will continue tomorrow :)

**Review by alittley — APPROVED** [review]
+1

File: platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
+# Durably Persist Self-Events Before Gossip
+
+---
+
+## Summary
+
+If a self-event is gossiped and the node is shutdown before the self-event is persisted to disk, then the node will
```
[the remark this answers is not in the document]
Comment (rbair23) [reply platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md:7]: Maybe we should have a link to the hashgraph whitepaper?

[the remark this answers is not in the document]
Comment (alittley) [reply platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md:7]: Are we going to put a link to the whitepaper in every proposal that has anything to do with consensus, events, transactions, etc.? I'm just wondering where we draw the line.

IMO, it should be ok to assume a base level of familiarity with fundamental platform concepts in these proposals.

[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md:7]: The intended audience for this writeup is somebody with a technical understanding of hashgraph algorithm and the core event pipeline as it currently exists. I'm happy to work with you to build up this understanding, but I don't think the proper place for transferring that knowledge is in this document.

File: platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+branch in the hashgraph on restart by building a new self-event on the same self-event parent it had previously used.
+In other words, _a node must only tell the world of a self-event **after** it has recorded this information!_
+
+This proposal ensures self-events are persisted to disk before they are gossiped. It simplifies the relationship
+between the consensus engine and the PCES writer, removing the need for the PCES sequencer, the `RoundDurabilityBuffer`,
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md:12]: I consider an understanding of the term "PCES" to be required background knowledge prior to reading this proposal. I think the proper level of description within a proposal should be one that a subject matter expert finds clear. With respect to becoming knowledgeable in this area, happy to assist. I just don't think the place to transfer that knowledge is in a design proposal.

File: platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md [hunk 4]
```diff
@@ -0,0 +1,202 @@
+# Durably Persist Self-Events Before Gossip
+
+---
+
+## Summary
+
+If a self-event is gossiped and the node is shutdown before the self-event is persisted to disk, then the node will
+branch in the hashgraph on restart by building a new self-event on the same self-event parent it had previously used.
```
Comment (jasperpotts) [inline platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md:8]: Is there a reason we cannot do a full sync with other nodes before the first event is created? That way you would know if other nodes know about an event created by you that you do not know about. If they do not then you are clear to use the last one you know about.

Comment (cody-littley) [reply platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md:8]: Unfortunately, such a thing is not mathematically plausible (it would be super convenient if it was!).

We must be functional as long as >2/3 of nodes (by weight) are participating in the protocol, since if we are not then a cohort of attackers representing <1/3 of the network can shut things down indefinitely. Even if we exclude the existence of malicious node operators, we fundamentally lack the ability to distinguish between a node that is permanently offline and a node that is slow. This is the fundamental nature of asynchronous networks (aka the "A" in "ABFT").

In summary, we can pause for a period of time in an attempt to give nodes a chance to participate (we already have a heuristic pause for this exact purpose). But unless we are willing to pause for an arbitrarily long period of time, there exists no algorithm that can guarantee that we always learn of our self events if we gossip them without writing them to disk.

File: platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+| HIPS               | N/A                                 |
+
+---
+
+## Purpose and Context
+
+Nodes in the network create new "self-events" on a regular basis. These self-events carry transactions that were
+submitted to the node, and advance the hashgraph. When a node creates a self-event, it must build upon the previous
+self-event and some "other parent" event (see the hashgraph algorithm for more details). If a node builds two
+self-events on the same self-event parent or if it creates an event with no self parent and there is a non-ancient self
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md:30]: sentence split up

File: platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+Today even honest nodes can accidentally branch (in practice the probability low but not unheard of). This happens
+when a node gossips a self-event and is shutdown before the event has been persisted to disk in the
+Pre-Consensus Event Stream (PCES). On restart, the PCES is replayed to populate the system with missing events,
+and other events that occurred since the node went down are retrieved via gossip. But there is a time between when
+the node starts up and sends a new self-event and when it receives its old self-event from other nodes. During this
+time, the node will build a new self-event on the same parent as the previous self-event (or with no self parent),
+causing a branch.
+
+This happens today most frequently on update boundaries when **every node** deterministically shutdown and then
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md:43]: change made

File: platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+start back up. In theory it can also happen when individual nodes are restarted.
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md:44]: change made

File: platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+### Requirements
+
+1. There must not be any noticeable impact in time-to-consensus.
```
[the remark this answers is not in the document]
Comment (cody-littley) [reply platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md:48]: change made

File: platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md (hunk 1 above)
Comment (cody-littley) [reply platform-sdk/docs/proposals/inlinePces/inlinePcesProposal.md:49]: collapsed the sub bullets

**Comment by cody-littley** [comment]
Closing this PR since we are not allowed to edit these without closing and reopening a new one.

**Comment by poulok** [comment]
>Closing this PR since we are not allowed to edit these without closing and reopening a new one.

Current agreement within the team is to allow immaterial modifications without requiring a new PR.

**Comment by cody-littley** [comment]
superseded by https://github.com/hashgraph/hedera-services/pull/14056
